--- ../src_base/common/net/minecraft/src/BlockTrapDoor.java
+++ ../src_work/common/net/minecraft/src/BlockTrapDoor.java
@@ -1,10 +1,14 @@
 package net.minecraft.src;
 
+import net.minecraftforge.common.ForgeDirection;
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
 
 public class BlockTrapDoor extends Block
 {
+    /** Set this to allow trapdoors to remain free-floating */
+    public static boolean disableValidation = false;
+
     protected BlockTrapDoor(int par1, Material par2Material)
     {
         super(par1, par2Material);
@@ -188,7 +192,7 @@
                 --var7;
             }
 
-            if (!isValidSupportBlock(par1World.getBlockId(var7, par3, var8)))
+            if (!(isValidSupportBlock(par1World.getBlockId(var7, par3, var8)) || par1World.isBlockSolidOnSide(var7, par3, var8, ForgeDirection.getOrientation((var6 & 3) + 2))))
             {
                 par1World.setBlockWithNotify(par2, par3, par4, 0);
                 this.dropBlockAsItem(par1World, par2, par3, par4, var6, 0);
@@ -248,6 +252,10 @@
      */
     public boolean canPlaceBlockOnSide(World par1World, int par2, int par3, int par4, int par5)
     {
+        if (disableValidation)
+        {
+            return true;
+        }
         if (par5 == 0)
         {
             return false;
@@ -278,7 +286,7 @@
                 --par2;
             }
 
-            return isValidSupportBlock(par1World.getBlockId(par2, par3, par4));
+            return isValidSupportBlock(par1World.getBlockId(par2, par3, par4)) || par1World.isBlockSolidOnSide(par2, par3, par4, ForgeDirection.UP);
         }
     }
 
@@ -293,6 +301,10 @@
      */
     private static boolean isValidSupportBlock(int par0)
     {
+        if (disableValidation)
+        {
+            return true;
+        }
         if (par0 <= 0)
         {
             return false;
@@ -303,4 +315,22 @@
             return var1 != null && var1.blockMaterial.isOpaque() && var1.renderAsNormalBlock() || var1 == Block.glowStone;
         }
     }
+    
+    @Override
+    public boolean rotateY(World world, int x, int y, int z, int amount)
+    {
+        amount = -amount;
+        int oldMeta = world.getBlockMetadata(x, y, z);
+        int pushed = (oldMeta & 4) >> 2;
+        int direction = oldMeta & 3;
+        boolean rotateNeg = amount < 0;
+        amount = Math.abs(amount) % 4;
+        if(rotateNeg)
+            amount = 4 - amount;
+        ForgeDirection dir = ForgeDirection.getOrientation(5 - direction);
+        for(int i = 0; i < amount; i++)
+            dir = dir.getRotCcwY();
+        world.setBlockMetadataWithNotify(x, y, z, (5 - dir.ordinal()) | (pushed << 2));
+        return true;
+    }
 }
