--- ../src_base/common/net/minecraft/src/BlockHalfSlab.java
+++ ../src_work/common/net/minecraft/src/BlockHalfSlab.java
@@ -166,4 +166,31 @@
      * Returns the slab block name with step type.
      */
     public abstract String getFullSlabName(int var1);
+    
+    @Override
+    public boolean rotateX(World world, int x, int y, int z, int amount)
+    {
+        if(isDoubleSlab)
+            return true;
+        int oldMeta = world.getBlockMetadata(x, y, z);
+        int type = oldMeta & 7;
+        int dir = (oldMeta & 8) >> 3;
+        amount = Math.abs(amount);
+        if(amount % 2 != 0)
+            return false;
+        if(amount % 4 == 0)
+            return true;
+        if(dir == 0)
+            dir = 1;
+        else if(dir == 1)
+            dir = 0;
+        world.setBlockMetadataWithNotify(x, y, z, dir << 3 | type);
+        return true;
+    }
+    
+    @Override
+    public boolean rotateZ(World world, int x, int y, int z, int amount)
+    {
+        return rotateX(world, x, y, z, amount);
+    }
 }
