--- ../src_base/minecraft_server/net/minecraft/src/NetLoginHandler.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/NetLoginHandler.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,10 +1,12 @@
 package net.minecraft.src;
 
 import java.io.IOException;
+import java.io.UnsupportedEncodingException;
 import java.net.Socket;
 import java.util.*;
 import java.util.logging.Logger;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.src.forge.network.MessageManager;
 
 public class NetLoginHandler extends NetHandler
 {
@@ -17,6 +19,7 @@
     private String username;
     private Packet1Login packet1login;
     private String serverId;
+    private boolean hasForge = false;
 
     public NetLoginHandler(MinecraftServer minecraftserver, Socket socket, String s)
     throws IOException
@@ -111,9 +114,12 @@
             entityplayermp.itemInWorldManager.setWorld((WorldServer)entityplayermp.worldObj);
             logger.info((new StringBuilder()).append(getUserAndIPString()).append(" logged in with entity id ").append(entityplayermp.entityId).append(" at (").append(entityplayermp.posX).append(", ").append(entityplayermp.posY).append(", ").append(entityplayermp.posZ).append(")").toString());
             WorldServer worldserver = mcServer.getWorldManager(entityplayermp.dimension);
-            ChunkCoordinates chunkcoordinates = worldserver.getSpawnPoint();
             entityplayermp.itemInWorldManager.func_35695_b(worldserver.getWorldInfo().getGameType());
-            NetServerHandler netserverhandler = new NetServerHandler(mcServer, netManager, entityplayermp);
+            final NetServerHandler netserverhandler = new NetServerHandler(mcServer, netManager, entityplayermp);
+            mcServer.networkServer.addPlayer(netserverhandler);
+            
+            /*
+            ChunkCoordinates chunkcoordinates = worldserver.getSpawnPoint();
             netserverhandler.sendPacket(new Packet1Login("", entityplayermp.entityId, worldserver.getSeed(), worldserver.getWorldInfo().getTerrainType(), entityplayermp.itemInWorldManager.getGameType(), (byte)worldserver.worldProvider.worldType, (byte)worldserver.difficultySetting, (byte)worldserver.worldHeight, (byte)mcServer.configManager.getMaxPlayers()));
             netserverhandler.sendPacket(new Packet6SpawnPosition(chunkcoordinates.posX, chunkcoordinates.posY, chunkcoordinates.posZ));
             mcServer.configManager.func_28170_a(entityplayermp, worldserver);
@@ -129,6 +135,44 @@
             }
 
             entityplayermp.func_20057_k();
+                
+            ModLoaderMp.HandleAllLogins(hasForge, entityplayermp);
+            */
+            try 
+            {
+                Packet250CustomPayload pkt = new Packet250CustomPayload();
+                pkt.channel = "REGISTER";
+                pkt.data    = "Forge".getBytes("UTF8");
+                pkt.length  = pkt.data.length;
+                netManager.addToSendQueue(pkt); //Register
+                
+                pkt = new Packet250CustomPayload();
+                pkt.channel = "Forge";
+                pkt.data    = new byte[]{0, 0, 0, 0, 0};
+                pkt.length  = pkt.data.length;
+                netManager.addToSendQueue(pkt); //Check for mods
+                (new Thread(){
+                    public void run(){
+                        try {
+                            this.sleep(1000);
+                            if (!netserverhandler.hasForge)
+                            {
+                                netserverhandler.kickPlayer("This server requires Minecraft Forge");
+                            }
+                        }
+                        catch (InterruptedException ex)
+                        {}
+                    }
+                }).start();
+            } 
+            catch (UnsupportedEncodingException e) 
+            {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
+                kickUser("Internal server error: " + e.getMessage());
+            }
+            
+            
         }
         finishedProcessing = true;
     }
